source/common/upstream/upstream_impl.cc

```diff
@@ -1469,11 +1470,12 @@ bool BaseDynamicClusterImpl::updateDynamicHostList(const HostVector& new_hosts,
         final_hosts.push_back(*i);
         updated_hosts[(*i)->address()->asString()] = *i;
         (*i)->healthFlagSet(Host::HealthFlag::PENDING_DYNAMIC_REMOVAL);
-        i = current_priority_hosts.erase(i);
+        *i = *(--erase_from_itr);
```

Comment:
in this instance I think you could just do lines 1470-1472 inside the predicate.

Reply:
This assuming the aforementioned refactoring? otherwise I don't understand.

Reply:
right exactly. since you'd be replacing this entire loop with a call to `Utility:: removeMatchingElements` you'd have to capture this extra work to be done on a match in the predicate.

I thought of another approach which is you could return a vector of items to remove, but I feel like doing it in the predicate is both less expensive and also more faithful to the flow here.